### PR TITLE
Support full yaml dsl #80

### LINKED ISSUE #80: Support full yaml dsl

### PR DESCRIPTION
**Pull request by lburgazzoli**
Fixes #80 

This is the first iteration for an YAMLesque DSL so that you can define routes as:

```yaml
- from:
    uri: "direct:start"
    steps:
     - choice:
         when:
           - simple: "${body.contains(\",\")}"
             steps:
               - to: "log:when-comma"
               - split:
                   tokenize: ","
                   steps:
                    - to: "log:when-semicolon"
           - simple: "${body.startsWith(\";\")}"
             steps:
               - to: "log:when-semicolon"
               - split:
                   tokenize: ";"
                   steps:
                    - to: "log:when-semicolon"
         otherwise:
           steps:
             - to: "log:otherwise"
     - to: "log:info"

- rest:
    verb: "post"
    uri: "/route"
    steps:
     - unmarshal:
         json:
             library: "Jackson"
             unmarshal-type: "my.Pojo"
     - to: "log:info"
```

The DSL can be extended by providing a "service" like file, see as example the custom step handler for knative here in the pr.

What's left:

- better documentation
- support for additional EIPs

The DSL support also a "flow style" where the inner steps for things such as `filter` or `split` can be omitted:

```yaml
- from:
    uri: "direct:flow"
    steps:
      - split:
          tokenizer: ","
      - to: "mock:flow"
```

### CODE REVIEW

File: camel-k-runtime-yaml/README.adoc [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+To make the DSL less verbose, the `expression` field can be omitted:
++
+[source, yaml]
+.Implicit Expression field
+----
+filter:
+    simple: "${in.header.continue} == true"
+----
++
+In general expression can be defined inline like in the examples above but in case you need provide more information, you can 'unroll' the expression definition and  configure any single parameter the expression defines.
```
Comment (davsclaus) [inline camel-k-runtime-yaml/README.adoc]: Double space in text

File: camel-k-runtime-yaml/README.adoc [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+----
+filter:
+    tokenize:
+      token: "<"
+      end-token: ">"
+----
+
+- *Data Format Aware Steps*
++
+Some EIP such as `set-body` and `marshal` supports the definition of a data formats through the `data-format` field:
```
Comment (davsclaus) [inline camel-k-runtime-yaml/README.adoc]: of a data formats -> of data formats

File: camel-k-runtime-yaml/README.adoc [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+        library: Jackson
+----
++
+To make the DSL less verbose, the `data-format` field can be omitted:
++
+[source, yaml]
+.Implicit Data Format field
+----
+set-body:
+    json:
```
Comment (davsclaus) [inline camel-k-runtime-yaml/README.adoc]: I think in Camel 3 we now use jackson as the default library

File: camel-k-runtime-yaml/src/main/java/org/apache/camel/k/yaml/Yaml.java [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+import com.fasterxml.jackson.dataformat.yaml.YAMLGenerator;
+import org.apache.camel.Exchange;
+import org.apache.camel.k.yaml.model.Definitions;
+import org.apache.camel.model.Block;
+import org.apache.camel.model.OptionalIdentifiedDefinition;
+import org.apache.camel.model.OtherAttributesAware;
+import org.apache.camel.model.ProcessorDefinition;
+
+public final class Yaml {
+    public final static ObjectMapper MAPPER = mapper();
```
Comment (davsclaus) [inline camel-k-runtime-yaml/src/main/java/org/apache/camel/k/yaml/Yaml.java]: I think there is a CS issue here

File: camel-k-runtime-yaml/src/main/java/org/apache/camel/k/yaml/Yaml.java [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+        );
+
+        return mapper;
+    }
+
+    public static JsonNode node(String name, JsonNode value) {
+        return Yaml.MAPPER.createObjectNode().set(name, value);
+    }
+
+    public abstract class ProcessorDefinitionMixIn<Type extends ProcessorDefinition<Type>>
```
Comment (davsclaus) [inline camel-k-runtime-yaml/src/main/java/org/apache/camel/k/yaml/Yaml.java]: Can you add some code comment or javadoc why we are doing this, so others can better understand

File: camel-k-runtime-yaml/src/main/java/org/apache/camel/k/yaml/YamlRoutesLoader.java [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+    @Override
+    public RouteBuilder load(CamelContext camelContext, Source source) throws Exception {
+        return builder(URIResolver.resolve(camelContext, source));
+    }
+
+    public static RouteBuilder builder(InputStream is) {
+        return new RouteBuilder() {
+            @Override
+            public void configure() throws Exception {
+                final CamelContext camelContext = getContext();
```
Comment (davsclaus) [inline camel-k-runtime-yaml/src/main/java/org/apache/camel/k/yaml/YamlRoutesLoader.java]: Ha this is clever to just do it in the configure method, then CamelContext just think this is a regular RouteBuilder class ;)

File: camel-k-runtime-yaml/src/main/java/org/apache/camel/k/yaml/converter/ChoiceConverter.java [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+        }
+
+        if (definition.otherwise != null) {
+            ConverterSupport.notNull(definition.otherwise.steps, "otherwise.steps");
+
+            ConverterSupport.convertSteps(
+                context,
+                choice.otherwise(),
+                definition.otherwise.steps)
+            ;
```
Comment (davsclaus) [inline camel-k-runtime-yaml/src/main/java/org/apache/camel/k/yaml/converter/ChoiceConverter.java]: Probably CS issues, can you run mvn -P sourcecheck

File: camel-k-runtime-yaml/src/main/java/org/apache/camel/k/yaml/model/Definitions.java [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+        return Collections.unmodifiableMap(definitions);
+    }
+
+    public static Map<String, Class<? extends ExpressionDefinition>> expressionDefinition() {
+        Map<String, Class<? extends ExpressionDefinition>> definitions = new HashMap<>();
+
+        definitions.put("constant", ConstantExpression.class);
+        definitions.put("header", HeaderExpression.class);
+        definitions.put("groovy", GroovyExpression.class);
+        definitions.put("hl4", Hl7TerserExpression.class);
```
Comment (valdar) [inline camel-k-runtime-yaml/src/main/java/org/apache/camel/k/yaml/model/Definitions.java]: should be hl7?

**Comment by davsclaus** [comment]
Looks really good. Only found cosmetic issues in the comments.

However I do think we should think about the naming of `converter` as it reminds too much about type converters from Camel. I would like to come up with a new name in the near future.

Really like the groovy testing stuff you do for all the EIPs.

**Comment by valdar** [comment]
+1 on the converter naming

**Comment by lburgazzoli** [comment]
@davsclaus @valdar yeah, the things have originally started as plain jackson converters so named according but it turned out it was not possible to fully marshal/unmarshal steps using standard things.

Any suggestion for a better naming is very welcome.

File: camel-k-runtime-yaml/src/main/java/org/apache/camel/k/yaml/converter/ChoiceConverter.java (hunk 7 above)
Comment (lburgazzoli) [reply camel-k-runtime-yaml/src/main/java/org/apache/camel/k/yaml/converter/ChoiceConverter.java]: don't think we have check style at the moment :)

File: camel-k-runtime-yaml/README.adoc (hunk 3 above)
Comment (lburgazzoli) [reply camel-k-runtime-yaml/README.adoc]: yes thius is just for documentation, will set to Gson so it is less confusing

**Comment by lburgazzoli** [comment]
ok, working on renaming things to StepParser

@valdar @davsclaus do you mind doing a second review ?

**Comment by davsclaus** [comment]
Yeah StepParser is better than Converter

File: camel-k-runtime-yaml/src/main/java/org/apache/camel/k/yaml/model/Definitions.java [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+    }
+
+    @SuppressWarnings("unchecked")
+    public static <T> Map<String, Class<? extends T>> loadDefinitionFromJaxbIndex(String packageName) {
+        Map<String, Class<? extends T>> definitions = new HashMap<>();
+        String path = "/" + packageName.replace(".", "/") + "/jaxb.index";
+
+        try (InputStream is = Definitions.class.getResourceAsStream(path)) {
+            for (String name : IOUtils.readLines(is, StandardCharsets.UTF_8)) {
+                name = name.trim();
```
Comment (valdar) [inline camel-k-runtime-yaml/src/main/java/org/apache/camel/k/yaml/model/Definitions.java]: might be better something like:
```
                name = name.trim();

	            if (!name.startsWith("#") && !name.isEmpty()) {
	                ClassLoader cl = Thread.currentThread().getContextClassLoader();
	                Class<?> clazz = cl.loadClass(packageName + "." + name);
	                XmlRootElement root = clazz.getAnnotation(XmlRootElement.class);

	                if (root != null) {
	                    definitions.put(root.name(), (Class<T>)clazz);
	                }
                }                
```

Comment (valdar) [reply camel-k-runtime-yaml/src/main/java/org/apache/camel/k/yaml/model/Definitions.java]: @lburgazzoli take a look

**Pull request merged**